### ISSUE TITLE
Can't sync against vaultwarden. Bitwarden client syncs OK.

### ISSUE THREAD

**Issue by BJReplay** [AChep/keyguard-app#84]
Keyguard reports:
> [!CAUTION]
> Couldn't sync the account.
> OK

Nothing particularly obvious from vaultwarden logs.

```
vaultwarden  | [2024-01-16 11:22:21.614][request][INFO] GET /api/sync?excludeDomains=true
vaultwarden  | [2024-01-16 11:22:22.157][response][INFO] (sync) GET /api/sync?<data..> => 200 OK
```

Running behind Caddy behind Cloudflare, but I have turned off Cloudflare just in case; no difference.

Let me know what I can provide in terms of logs to assist in debugging.

Only difference I can spot is 

`GET /api/sync?excludeDomains=true` from KeyGuard vs `GET /api/sync` from Bitwarden clients.
Response from vaultwarden for both is ` (sync) GET /api/sync?<data..> => 200 OK`

Just in case the cause was the `?excludeDomains=true` query parameters, set up a URL rewrite to remove it, but didn't make any difference.

The failure is instantaneous; it occurs at the beginning of sync.

**Comment by yurividal** [comment]
It works fine with my 2 vaultwarden instances. In my case both are behind nginx

**Comment by BJReplay** [comment]
I wonder if it is related to having multiple organisations (and/or collections, associated with those organisations)?

**Comment by BJReplay** [comment]
OK, so it's not Caddy, and not the server.

A new account, not yet joined to any other organisations or with any logins or folders, logs in and syncs.

I'll set up 2FA next, and then try importing all non-organisation items, first, and if that works, then join organisations one by one (I have three), and see what breaks.

**Comment by BJReplay** [comment]
OK, so it's not 2FA, or any of the items in my vault.

Next stop, joining organisations...

And if that works, I'm going to have to assume that there was something funky with my original account...

**Comment by BJReplay** [comment]
OK, I (almost) give up.

I have imported all my own vault, and added myself to all three organisations, and everything syncs.

Final test will be to log out of the new user account (now that it is fully set up), and log in again, and see if it syncs, now that it is fully set up.

**Comment by BJReplay** [comment]
And it all works.

I think there is something funky with my first account.  Feel free to close this, or keep it open; if you have some way of getting more logging out of the sync attempts, I can reproduce the error every time I attempt to log in with my main account.  I'll leave it open for now, but if you close it, I'll understand why.  If you leave it open, with instructions on gathering logs, I'll do all I can to assist.

**Comment by AChep** [comment]
@BJReplay any chance you could record ADB logcat and then try to connect to the OG account again? The log might contain sensitive information, so please sanitize it if needed and send it to my email (artemchep [at] gmail.com) so less people see if you miss something accidentally.

**Comment by BJReplay** [comment]
> could record ADB logcat

reviewing now... didn't find what I was looking for in my first pass at sanitisation, but did find

`System.err: Caused by: fw.c: Field 'error' is required for type with serial name 'com.artemchep.keyguard.provider.bitwarden.entity.ErrorEntity', but it was missing at path: $`

**Comment by BJReplay** [comment]
I'm going to spend a bit more time looking for stuff I need to sanitise, but at first glance, it's a missing date for a password history element (which might explain why the imported account didn't have this problem):

```
1-19 16:49:00.614 15168 15797 W System.err: dr.f: Illegal input: Field 'LastUsedDate' is required for type with serial name 'com.artemchep.keyguard.provider.bitwarden.entity.PasswordHistoryEntity', but it was missing at path: $.Ciphers[3].PasswordHistory[0]
01-19 16:49:00.614 15168 15797 W System.err: 	at er.h.a(Unknown Source:310)
01-19 16:49:00.614 15168 15797 W System.err: 	at dr.b.b(Unknown Source:82)
01-19 16:49:00.614 15168 15797 W System.err: 	at vv.k.c(Unknown Source:190)
01-19 16:49:00.614 15168 15797 W System.err: 	at dr.c.c(Unknown Source:24)
01-19 16:49:00.614 15168 15797 W System.err: 	at de.j7.y(Unknown Source:72)
01-19 16:49:00.614 15168 15797 W System.err: 	at fa.i.m(Unknown Source:81)
01-19 16:49:00.614 15168 15797 W System.err: 	at tq.f.b(Unknown Source:245)
01-19 16:49:00.614 15168 15797 W System.err: 	at kq.a.j(Unknown Source:600)
01-19 16:49:00.614 15168 15797 W System.err: 	at kq.a.n(Unknown Source:22)
01-19 16:49:00.614 15168 15797 W System.err: 	at kq.a.invoke(Unknown Source:70)
01-19 16:49:00.614 15168 15797 W System.err: 	at ir.k.g(Unknown Source:35)
01-19 16:49:00.614 15168 15797 W System.err: 	at ir.k.d(Unknown Source:27)
01-19 16:49:00.614 15168 15797 W System.err: 	at kq.b.j(Unknown Source:48)
01-19 16:49:00.614 15168 15797 W System.err: 	at kq.b.invoke(Unknown Source:24)
01-19 16:49:00.614 15168 15797 W System.err: 	at ir.k.g(Unknown Source:35)
01-19 16:49:00.614 15168 15797 W System.err: 	at ir.k.d(Unknown Source:27)
01-19 16:49:00.614 15168 15797 W System.err: 	at kq.a.j(Unknown Source:326)
01-19 16:49:00.614 15168 15797 W System.err: 	at kq.a.n(Unknown Source:37)
01-19 16:49:00.614 15168 15797 W System.err: 	at kq.a.invoke(Unknown Source:59)
01-19 16:49:00.614 15168 15797 W System.err: 	at ir.k.g(Unknown Source:35)
01-19 16:49:00.614 15168 15797 W System.err: 	at ir.k.d(Unknown Source:27)
01-19 16:49:00.614 15168 15797 W System.err: 	at ir.k.f(Unknown Source:7)
01-19 16:49:00.614 15168 15797 W System.err: 	at qq.m.j(Unknown Source:63)
01-19 16:49:00.614 15168 15797 W System.err: 	at qq.m.invoke(Unknown Source:26)
01-19 16:49:00.614 15168 15797 W System.err: 	at ir.k.g(Unknown Source:35)
01-19 16:49:00.614 15168 15797 W System.err: 	at ir.k.d(Unknown Source:27)
01-19 16:49:00.614 15168 15797 W System.err: 	at ir.k.a(Unknown Source:23)
01-19 16:49:00.614 15168 15797 W System.err: 	at ir.d.a(Unknown Source:189)
01-19 16:49:00.614 15168 15797 W System.err: 	at lq.b.a(Unknown Source:165)
01-19 16:49:00.614 15168 15797 W System.err: 	at ug.q.q(Unknown Source:284)
01-19 16:49:00.614 15168 15797 W System.err: 	at ug.o.j(Unknown Source:10)
01-19 16:49:00.614 15168 15797 W System.err: 	at ms.a.q(Unknown Source:8)
01-19 16:49:00.614 15168 15797 W System.err: 	at sv.n0.run(Unknown Source:107)
01-19 16:49:00.614 15168 15797 W System.err: 	at yv.a.run(Unknown Source:98)
01-19 16:49:00.615 15168 15797 W System.err: Caused by: fw.c: Field 'LastUsedDate' is required for type with serial name 'com.artemchep.keyguard.provider.bitwarden.entity.PasswordHistoryEntity', but it was missing at path: $.Ciphers[3].PasswordHistory[0]
01-19 16:49:00.615 15168 15797 W System.err: 	at kw.v.l(Unknown Source:158)
01-19 16:49:00.615 15168 15797 W System.err: 	at fl.l1.E(Unknown Source:10)
01-19 16:49:00.615 15168 15797 W System.err: 	at kw.v.E(Unknown Source:49)
01-19 16:49:00.615 15168 15797 W System.err: 	at iw.s.k(Unknown Source:7)
01-19 16:49:00.615 15168 15797 W System.err: 	at iw.a.j(Unknown Source:37)
01-19 16:49:00.615 15168 15797 W System.err: 	at iw.a.d(Unknown Source:5)
01-19 16:49:00.615 15168 15797 W System.err: 	at kw.v.l(Unknown Source:102)
01-19 16:49:00.615 15168 15797 W System.err: 	at fl.l1.f(Unknown Source:32)
01-19 16:49:00.615 15168 15797 W System.err: 	at xg.p.d(Unknown Source:123)
01-19 16:49:00.615 15168 15797 W System.err: 	at kw.v.l(Unknown Source:102)
01-19 16:49:00.615 15168 15797 W System.err: 	at fl.l1.E(Unknown Source:10)
01-19 16:49:00.615 15168 15797 W System.err: 	at kw.v.E(Unknown Source:49)
01-19 16:49:00.615 15168 15797 W System.err: 	at iw.s.k(Unknown Source:7)
01-19 16:49:00.615 15168 15797 W System.err: 	at iw.a.j(Unknown Source:37)
01-19 16:49:00.615 15168 15797 W System.err: 	at iw.a.d(Unknown Source:5)
01-19 16:49:00.615 15168 15797 W System.err: 	at kw.v.l(Unknown Source:102)
01-19 16:49:00.615 15168 15797 W System.err: 	at fl.l1.f(Unknown Source:32)
01-19 16:49:00.615 15168 15797 W System.err: 	at vg.b.d(Unknown Source:108)
01-19 16:49:00.615 15168 15797 W System.err: 	at kw.v.l(Unknown Source:102)
01-19 16:49:00.615 15168 15797 W System.err: 	at jw.b.a(Unknown Source:30)
01-19 16:49:00.615 15168 15797 W System.err: 	at er.h.a(Unknown Source:201)
01-19 16:49:00.615 15168 15797 W System.err: 	... 33 more
01-19 16:49:00.615 15168 15797 W System.err: Caused by: fw.c: Field 'LastUsedDate' is required for type with serial name 'com.artemchep.keyguard.provider.bitwarden.entity.PasswordHistoryEntity', but it was missing
01-19 16:49:00.615 15168 15797 W System.err: 	at i7.f.A(Unknown Source:105)
01-19 16:49:00.615 15168 15797 W System.err: 	at xg.y2.<init>(Unknown Source:15)
01-19 16:49:00.615 15168 15797 W System.err: 	at xg.w2.d(Unknown Source:64)
01-19 16:49:00.615 15168 15797 W System.err: 	at kw.v.l(Unknown Source:102)
01-19 16:49:00.615 15168 15797 W System.err: 	... 53 more

```

**Comment by AChep** [comment]
> I'm going to spend a bit more time looking for stuff I need to sanitise, but at first glance, it's a missing date for a password history element (which might explain why the imported account didn't have this problem):
> 
> ```
> 1-19 16:49:00.614 15168 15797 W System.err: dr.f: Illegal input: Field 'LastUsedDate' is required for type with serial name 'com.artemchep.keyguard.provider.bitwarden.entity.PasswordHistoryEntity', but it was missing at path: $.Ciphers[3].PasswordHistory[0]
> 01-19 16:49:00.614 15168 15797 W System.err: 	at er.h.a(Unknown Source:310)
> 01-19 16:49:00.614 15168 15797 W System.err: 	at dr.b.b(Unknown Source:82)
> 01-19 16:49:00.614 15168 15797 W System.err: 	at vv.k.c(Unknown Source:190)
> 01-19 16:49:00.614 15168 15797 W System.err: 	at dr.c.c(Unknown Source:24)
> 01-19 16:49:00.614 15168 15797 W System.err: 	at de.j7.y(Unknown Source:72)
> 01-19 16:49:00.614 15168 15797 W System.err: 	at fa.i.m(Unknown Source:81)
> 01-19 16:49:00.614 15168 15797 W System.err: 	at tq.f.b(Unknown Source:245)
> 01-19 16:49:00.614 15168 15797 W System.err: 	at kq.a.j(Unknown Source:600)
> 01-19 16:49:00.614 15168 15797 W System.err: 	at kq.a.n(Unknown Source:22)
> 01-19 16:49:00.614 15168 15797 W System.err: 	at kq.a.invoke(Unknown Source:70)
> 01-19 16:49:00.614 15168 15797 W System.err: 	at ir.k.g(Unknown Source:35)
> 01-19 16:49:00.614 15168 15797 W System.err: 	at ir.k.d(Unknown Source:27)
> 01-19 16:49:00.614 15168 15797 W System.err: 	at kq.b.j(Unknown Source:48)
> 01-19 16:49:00.614 15168 15797 W System.err: 	at kq.b.invoke(Unknown Source:24)
> 01-19 16:49:00.614 15168 15797 W System.err: 	at ir.k.g(Unknown Source:35)
> 01-19 16:49:00.614 15168 15797 W System.err: 	at ir.k.d(Unknown Source:27)
> 01-19 16:49:00.614 15168 15797 W System.err: 	at kq.a.j(Unknown Source:326)
> 01-19 16:49:00.614 15168 15797 W System.err: 	at kq.a.n(Unknown Source:37)
> 01-19 16:49:00.614 15168 15797 W System.err: 	at kq.a.invoke(Unknown Source:59)
> 01-19 16:49:00.614 15168 15797 W System.err: 	at ir.k.g(Unknown Source:35)
> 01-19 16:49:00.614 15168 15797 W System.err: 	at ir.k.d(Unknown Source:27)
> 01-19 16:49:00.614 15168 15797 W System.err: 	at ir.k.f(Unknown Source:7)
> 01-19 16:49:00.614 15168 15797 W System.err: 	at qq.m.j(Unknown Source:63)
> 01-19 16:49:00.614 15168 15797 W System.err: 	at qq.m.invoke(Unknown Source:26)
> 01-19 16:49:00.614 15168 15797 W System.err: 	at ir.k.g(Unknown Source:35)
> 01-19 16:49:00.614 15168 15797 W System.err: 	at ir.k.d(Unknown Source:27)
> 01-19 16:49:00.614 15168 15797 W System.err: 	at ir.k.a(Unknown Source:23)
> 01-19 16:49:00.614 15168 15797 W System.err: 	at ir.d.a(Unknown Source:189)
> 01-19 16:49:00.614 15168 15797 W System.err: 	at lq.b.a(Unknown Source:165)
> 01-19 16:49:00.614 15168 15797 W System.err: 	at ug.q.q(Unknown Source:284)
> 01-19 16:49:00.614 15168 15797 W System.err: 	at ug.o.j(Unknown Source:10)
> 01-19 16:49:00.614 15168 15797 W System.err: 	at ms.a.q(Unknown Source:8)
> 01-19 16:49:00.614 15168 15797 W System.err: 	at sv.n0.run(Unknown Source:107)
> 01-19 16:49:00.614 15168 15797 W System.err: 	at yv.a.run(Unknown Source:98)
> 01-19 16:49:00.615 15168 15797 W System.err: Caused by: fw.c: Field 'LastUsedDate' is required for type with serial name 'com.artemchep.keyguard.provider.bitwarden.entity.PasswordHistoryEntity', but it was missing at path: $.Ciphers[3].PasswordHistory[0]
> 01-19 16:49:00.615 15168 15797 W System.err: 	at kw.v.l(Unknown Source:158)
> 01-19 16:49:00.615 15168 15797 W System.err: 	at fl.l1.E(Unknown Source:10)
> 01-19 16:49:00.615 15168 15797 W System.err: 	at kw.v.E(Unknown Source:49)
> 01-19 16:49:00.615 15168 15797 W System.err: 	at iw.s.k(Unknown Source:7)
> 01-19 16:49:00.615 15168 15797 W System.err: 	at iw.a.j(Unknown Source:37)
> 01-19 16:49:00.615 15168 15797 W System.err: 	at iw.a.d(Unknown Source:5)
> 01-19 16:49:00.615 15168 15797 W System.err: 	at kw.v.l(Unknown Source:102)
> 01-19 16:49:00.615 15168 15797 W System.err: 	at fl.l1.f(Unknown Source:32)
> 01-19 16:49:00.615 15168 15797 W System.err: 	at xg.p.d(Unknown Source:123)
> 01-19 16:49:00.615 15168 15797 W System.err: 	at kw.v.l(Unknown Source:102)
> 01-19 16:49:00.615 15168 15797 W System.err: 	at fl.l1.E(Unknown Source:10)
> 01-19 16:49:00.615 15168 15797 W System.err: 	at kw.v.E(Unknown Source:49)
> 01-19 16:49:00.615 15168 15797 W System.err: 	at iw.s.k(Unknown Source:7)
> 01-19 16:49:00.615 15168 15797 W System.err: 	at iw.a.j(Unknown Source:37)
> 01-19 16:49:00.615 15168 15797 W System.err: 	at iw.a.d(Unknown Source:5)
> 01-19 16:49:00.615 15168 15797 W System.err: 	at kw.v.l(Unknown Source:102)
> 01-19 16:49:00.615 15168 15797 W System.err: 	at fl.l1.f(Unknown Source:32)
> 01-19 16:49:00.615 15168 15797 W System.err: 	at vg.b.d(Unknown Source:108)
> 01-19 16:49:00.615 15168 15797 W System.err: 	at kw.v.l(Unknown Source:102)
> 01-19 16:49:00.615 15168 15797 W System.err: 	at jw.b.a(Unknown Source:30)
> 01-19 16:49:00.615 15168 15797 W System.err: 	at er.h.a(Unknown Source:201)
> 01-19 16:49:00.615 15168 15797 W System.err: 	... 33 more
> 01-19 16:49:00.615 15168 15797 W System.err: Caused by: fw.c: Field 'LastUsedDate' is required for type with serial name 'com.artemchep.keyguard.provider.bitwarden.entity.PasswordHistoryEntity', but it was missing
> 01-19 16:49:00.615 15168 15797 W System.err: 	at i7.f.A(Unknown Source:105)
> 01-19 16:49:00.615 15168 15797 W System.err: 	at xg.y2.<init>(Unknown Source:15)
> 01-19 16:49:00.615 15168 15797 W System.err: 	at xg.w2.d(Unknown Source:64)
> 01-19 16:49:00.615 15168 15797 W System.err: 	at kw.v.l(Unknown Source:102)
> 01-19 16:49:00.615 15168 15797 W System.err: 	... 53 more
> 
> ```

It should be it! I will make it optional on my end.

**Comment by BJReplay** [comment]
Look for line 355538 when you get the dump from me.

**Comment by AChep** [comment]
Should be fixed in the next build, that should be there tomorrow.

**Comment by AChep** [comment]
@BJReplay you can test if the fix worked here https://github.com/AChep/keyguard-app/actions/runs/7581328386

**Comment by BJReplay** [comment]
> @BJReplay you can test if the fix worked here https://github.com/AChep/keyguard-app/actions/runs/7581328386

Won't be able to properly test until it appears in the Google Play store, sorry - I'm in the Advanced Protection Programme so can't install APKs on this phone.  I have an old phone that I can play with that's rooted running under another account, but it isn't running android 14, so I'll give it a go there (and at least test the account aspect) and let you know.  Let me know once it's pushed to the store and I'll test.

**Comment by BJReplay** [comment]
That build fixes it.  Also tested with my self-hosted SimpleLogin! Nice.

**Comment by BJReplay** [comment]
```python
from bitwardentools import Client
from bitwardentools import CipherType

def api_key(loginpayload):
    loginpayload.update(
        {
            "client_id": client_id,
            "client_secret": client_secret,
            "scope": "api",
            "grant_type": "client_credentials"
        }
    )
    return loginpayload

vault = "https://vault.domain.tld"
user = "user@domain.tld"
password = "password"
client_id = "client_id"
client_secret = "client_secret"

myclient = Client(vault,user,password,authentication_cb=api_key)

all_ciphers_ids = myclient.get_ciphers()['id']
for id, login in all_ciphers_ids.items():
    if login.type == CipherType.Login: # Only process logins
        if login.passwordHistory is not None:
            print()
            print(f"Found Login with History '{login.name}' for {login.login['username']}") #log output
            for i, pwHist in enumerate(login.passwordHistory):
                print(f"  lastUsedDate  '{pwHist['lastUsedDate']}'")
myclient.bust_cache
```

> Found Login with History 'turkishairlines.com' for 9999999999
>   lastUsedDate  '2023-12-22T14:09:17.530Z'
>   lastUsedDate  '2023-12-22T14:09:05.650Z'
>   lastUsedDate  '2023-12-22T14:08:53.313Z'
>   lastUsedDate  '1970-01-01T00:00:00.000Z'
> 
> Found Login with History 'oracle.com' for name.surname@gmail.com
>   lastUsedDate  '2023-12-22T14:34:01.903Z'
>   lastUsedDate  '2023-12-22T14:33:40.159Z'
>   lastUsedDate  '1970-01-01T00:00:00.000Z'
>   lastUsedDate  '1970-01-01T00:00:00.000Z'
> 
> Found Login with History 'oracle.com - Backup' for backup.account@gmail.com
>   lastUsedDate  '2023-12-22T14:34:25.153Z'
>   lastUsedDate  '2023-12-22T14:34:14.839Z'
>   lastUsedDate  '1970-01-01T00:00:00.000Z'

And many others without invalid dates...

Note the last entries in each list.

It's not a null date as returned after interpretation, but near enough to it.

In the bitwarden UI, these display as 1 Jan 1970.  Its a zero unix timestamp.

**Comment by AChep** [comment]
@BJReplay  Is it not working with the updated build when the timestamp is 0?

**Comment by BJReplay** [comment]
Yes, it is working on my test phone (Android 13)  I just went looking for the culprit ciphers since I had bitwardentools installed, and was interested in finding the dodgy ciphers.

I just rotated the passwords in those accounts enough times to push the oldest passwords out and I can log in now with the existing play store build on my locked down Android 14 phone.

I shouldn't have re-opened the issue, sorry, it was more FYI.
